receiver/dcgmreceiver/client.go

```diff
@@ -83,14 +81,26 @@ func newClient(config *Config, logger *zap.Logger) (*dcgmClient, error) {
 	for _, f := range unavailableFields {
 		logger.Sugar().Warnf("Field '%s' is not supported. Metric '%s' will not be collected", dcgmIDToName[f], dcgmNameToMetricName[dcgmIDToName[f]])
 	}
+	// If there is no metrics to watch after filtering the user configuration
+	// with device supported fields, return an client with an empty device list
+	if len(enabledFields) == 0 {
+		return &dcgmClient{
+			logger:                         logger.Sugar(),
+			enabledFieldIDs:                enabledFields,
+			enabledFieldGroup:              dcgm.FieldHandle{},
+			deviceIndices:                  make([]uint, 0),
+			devicesModelName:               make([]string, 0),
+			devicesUUID:                    make([]string, 0),
+			deviceMetricToFailedQueryCount: make(map[string]uint64, 0),
+		}, nil
```

Comment:
This seems like code duplication, and a bunch of unnecessary work to discover the available devices, etc, only to ignore them here. Unless I'm missing something, both `requestedFieldIDs` and `supportedFieldIDs` don't depend on any of the other computed values, and thus can be queried first, so this could be short-circuited way earlier.

To be clear, here's what I'm suggesting:
```
func newClient(config *Config, logger *zap.Logger) (*dcgmClient, error) {
	deviceIndices := make([]uint, 0)
	names := make([]string, 0)
	UUIDs := make([]string, 0)
	enabledFieldGroup := dcgm.FieldHandle{},
	requestedFieldIDs := discoverRequestedFieldIDs(config)
	supportedFieldIDs, err := getAllSupportedFields()
	…
	enabledFields, unavailableFields := filterSupportedFields(requestedFieldIDs, supportedFieldIDs)
	…
	if len(enabledFields) != 0 {
		deviceIndices, names, UUIDs, err = discoverDevices(logger)
		…
		deviceGroup, err = createDeviceGroup(logger, deviceIndices)
		…
		enabledFieldGroup, err = setWatchesOnEnabledFields(config, logger, deviceGroup, enabledFields)
		…
	}
	return &dcgmClient{
		logger:                         logger.Sugar(),
		enabledFieldIDs:                enabledFields,
		enabledFieldGroup:              enabledFieldGroup,
		deviceIndices:                  deviceIndices,
		devicesModelName:               names,
		devicesUUID:                    UUIDs,
		deviceMetricToFailedQueryCount: make(map[string]uint64),
	}, nil
}
```

Reply:
Done! Much cleaner right now. Thanks!